File: backend/app/services/analytics_service.py
"""Analytics service for the admin dashboard & analytics page."""
from __future__ import annotations

from collections import defaultdict
from datetime import datetime, timedelta
from typing import List, Optional

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.models import (
    District,
    Image,
    InfrastructureType,
    PriorityScore,
    Report,
    User,
    Verification,
    SEVERITY_CRITICAL,
    SEVERITY_HIGH,
    SEVERITY_LOW,
    SEVERITY_MODERATE,
    STATUS_REPORTED,
    STATUS_VERIFIED,
    STATUS_RESOLVED,
)
from app.schemas import (
    CategoryDistributionItem,
    DashboardSummary,
    DistrictAnalyticsItem,
    MonthlyTrendItem,
    SeverityDistributionItem,
)


SEVERITY_ORDER = [SEVERITY_LOW, SEVERITY_MODERATE, SEVERITY_HIGH, SEVERITY_CRITICAL]


def _parse_dt(val) -> Optional[datetime]:
    if val is None:
        return None
    if isinstance(val, datetime):
        return val
    try:
        # Strip trailing fraction/tz if needed
        clean = str(val).replace("Z", "").split("+")[0]
        if "." in clean:
            clean = clean.split(".")[0]
        return datetime.fromisoformat(clean)
    except Exception:
        return None


def dashboard_summary(db: Session) -> DashboardSummary:
    total = db.execute(select(func.count(Report.id))).scalar_one()
    pending = db.execute(select(func.count(Report.id)).where(Report.status == STATUS_REPORTED)).scalar_one()
    verified = db.execute(select(func.count(Report.id)).where(Report.status == STATUS_VERIFIED)).scalar_one()
    resolved = db.execute(select(func.count(Report.id)).where(Report.status == STATUS_RESOLVED)).scalar_one()
    critical = db.execute(
        select(func.count(Report.id)).where(
            (Report.ai_severity == SEVERITY_CRITICAL) | (Report.final_severity == SEVERITY_CRITICAL)
        )
    ).scalar_one()
    users = db.execute(select(func.count(User.id))).scalar_one()
    verifications = db.execute(select(func.count(Verification.id))).scalar_one()

    # Avg response time (hours) = resolved_at - created_at for resolved reports
    resolved_reports = db.execute(
        select(Report.created_at, Report.resolved_at)
        .where(Report.status == STATUS_RESOLVED, Report.resolved_at.isnot(None))
    ).all()
    durations = []
    for c_at_raw, r_at_raw in resolved_reports:
        c_at = _parse_dt(c_at_raw)
        r_at = _parse_dt(r_at_raw)
        if c_at and r_at and r_at >= c_at:
            durations.append((r_at - c_at).total_seconds() / 3600.0)
    avg_response = (sum(durations) / len(durations)) if durations else None
    response_rate = (resolved / total * 100.0) if total else 0.0

    return DashboardSummary(
        total_reports=total,
        pending_reports=pending,
        verified_reports=verified,
        resolved_reports=resolved,
        critical_incidents=critical,
        total_users=users,
        total_verifications=verifications,
        avg_response_time_hours=round(avg_response, 2) if avg_response else None,
        response_rate=round(response_rate, 2),
    )


def severity_distribution(db: Session) -> List[SeverityDistributionItem]:
    rows = db.execute(
        select(Report.ai_severity, func.count(Report.id)).group_by(Report.ai_severity)
    ).all()
    counts = {r[0] or "Unassessed": r[1] for r in rows}
    total = sum(counts.values()) or 1
    items = []
    for sev in SEVERITY_ORDER + ["Unassessed"]:
        c = counts.get(sev, 0)
        items.append(SeverityDistributionItem(
            severity=sev, count=c, percentage=round(c / total * 100.0, 2)
        ))
    return items


def category_distribution(db: Session) -> List[CategoryDistributionItem]:
    categories = db.execute(select(InfrastructureType)).scalars().all()
    items = []
    for cat in categories:
        reports = db.execute(
            select(Report).where(Report.infrastructure_type_id == cat.id)
        ).scalars().all()
        critical_count = sum(
            1 for r in reports
            if (r.ai_severity == SEVERITY_CRITICAL or r.final_severity == SEVERITY_CRITICAL)
        )
        items.append(CategoryDistributionItem(
            category=cat.name,
            count=len(reports),
            critical_count=critical_count,
        ))
    return items


def monthly_trend(db: Session, months: int = 6) -> List[MonthlyTrendItem]:
    since = datetime.utcnow() - timedelta(days=months * 30)
    reports = db.execute(select(Report.created_at, Report.status)).all()
    data = defaultdict(lambda: {"reports": 0, "resolved": 0})
    for created_at_raw, status in reports:
        dt = _parse_dt(created_at_raw)
        if dt and dt >= since:
            m = dt.strftime("%Y-%m")
            data[m]["reports"] += 1
            if status == STATUS_RESOLVED:
                data[m]["resolved"] += 1
    items = [
        MonthlyTrendItem(month=m, reports=counts["reports"], resolved=counts["resolved"])
        for m, counts in sorted(data.items())
    ]
    return items


def district_analytics(db: Session) -> List[DistrictAnalyticsItem]:
    districts = db.execute(select(District)).scalars().all()
    items = []
    for d in districts:
        reports = db.execute(
            select(Report).where(Report.district_id == d.id)
        ).scalars().all()
        critical_count = sum(
            1 for r in reports
            if (r.ai_severity == SEVERITY_CRITICAL or r.final_severity == SEVERITY_CRITICAL)
        )
        resolved_count = sum(1 for r in reports if r.status == STATUS_RESOLVED)
        scores = []
        for r in reports:
            ps = db.execute(
                select(PriorityScore)
                .where(PriorityScore.report_id == r.id)
                .order_by(PriorityScore.created_at.desc())
                .limit(1)
            ).scalar_one_or_none()
            if ps and ps.score is not None:
                scores.append(ps.score)
        avg_priority = round(sum(scores) / len(scores), 2) if scores else 0.0
        items.append(DistrictAnalyticsItem(
            district=d.name,
            reports=len(reports),
            critical=critical_count,
            resolved=resolved_count,
            avg_priority=avg_priority,
        ))
    return items


def response_time_analytics(db: Session) -> dict:
    resolved_reports = db.execute(
        select(Report.created_at, Report.resolved_at)
        .where(Report.status == STATUS_RESOLVED, Report.resolved_at.isnot(None))
    ).all()
    durations = [
        (r_at - c_at).total_seconds() / 3600.0
        for c_at, r_at in resolved_reports
        if c_at and r_at and r_at >= c_at
    ]
    if not durations:
        return {"avg_hours": None, "min_hours": None, "max_hours": None}
    return {
        "avg_hours": round(sum(durations) / len(durations), 2),
        "min_hours": round(min(durations), 2),
        "max_hours": round(max(durations), 2),
    }


def repeat_incidents(db: Session, threshold_km: float = 0.5) -> List[dict]:
    """Detect reports within threshold_km of each other for the same infra type."""
    reports = db.execute(
        select(Report).where(Report.status != "Rejected")
    ).scalars().all()
    clusters = []
    used = set()
    for i, r1 in enumerate(reports):
        if r1.id in used:
            continue
        group = [r1]
        for r2 in reports[i + 1:]:
            if r2.id in used or r2.infrastructure_type_id != r1.infrastructure_type_id:
                continue
            # Haversine approx
            d = _haversine(r1.latitude, r1.longitude, r2.latitude, r2.longitude)
            if d <= threshold_km:
                group.append(r2)
                used.add(r2.id)
        if len(group) > 1:
            clusters.append({
                "infrastructure_type_id": r1.infrastructure_type_id,
                "center": {"lat": r1.latitude, "lng": r1.longitude},
                "count": len(group),
                "report_ids": [r.id for r in group],
            })
            used.add(r1.id)
    return clusters


def _haversine(lat1, lon1, lat2, lon2) -> float:
    import math
    R = 6371.0
    phi1, phi2 = math.radians(lat1), math.radians(lat2)
    dphi = math.radians(lat2 - lat1)
    dlam = math.radians(lon2 - lon1)
    a = math.sin(dphi / 2) ** 2 + math.cos(phi1) * math.cos(phi2) * math.sin(dlam / 2) ** 2
    return 2 * R * math.asin(math.sqrt(a))


def citizen_participation(db: Session) -> dict:
    total_users = db.execute(select(func.count(User.id))).scalar_one()
    active_users = db.execute(
        select(func.count(func.distinct(Report.user_id)))
    ).scalar_one()
    verifiers = db.execute(select(func.count(func.distinct(Verification.user_id)))).scalar_one()
    avg_verifications_per_report = db.execute(
        select(func.avg(Report.verification_count))
    ).scalar_one()
    return {
        "total_citizens": total_users,
        "citizens_who_reported": active_users,
        "citizens_who_verified": verifiers,
        "avg_verifications_per_report": round(float(avg_verifications_per_report), 2)
        if avg_verifications_per_report else 0.0,
    }

File: backend-java/src/main/java/com/infraguard/service/AnalyticsService.java
package com.infraguard.service;

import com.infraguard.dto.analytics.*;
import com.infraguard.entity.*;
import com.infraguard.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class AnalyticsService {

    private final ReportRepository reportRepository;
    private final UserRepository userRepository;
    private final VerificationRepository verificationRepository;
    private final InfrastructureTypeRepository infraTypeRepository;
    private final DistrictRepository districtRepository;
    private final PriorityScoreRepository priorityScoreRepository;

    @Transactional(readOnly = true)
    public DashboardSummary dashboardSummary() {
        long total = reportRepository.count();
        long pending = reportRepository.countByStatus("Reported");
        long verified = reportRepository.countByStatus("Verified");
        long resolved = reportRepository.countByStatus("Resolved");
        long critical = reportRepository.countCritical();
        long users = userRepository.count();
        long verifications = verificationRepository.count();
        Double avgResp = reportRepository.avgResponseTimeHours();
        double responseRate = total > 0 ? (resolved * 100.0 / total) : 0.0;

        return DashboardSummary.builder()
            .totalReports(total)
            .pendingReports(pending)
            .verifiedReports(verified)
            .resolvedReports(resolved)
            .criticalIncidents(critical)
            .totalUsers(users)
            .totalVerifications(verifications)
            .avgResponseTimeHours(avgResp != null ? Math.round(avgResp * 100.0) / 100.0 : null)
            .responseRate(Math.round(responseRate * 100.0) / 100.0)
            .build();
    }

    @Transactional(readOnly = true)
    public List<SeverityDistributionItem> severityDistribution() {
        String[] severities = {"Low", "Moderate", "High", "Critical", "Unassessed"};
        long total = reportRepository.count() + 1;
        List<SeverityDistributionItem> result = new ArrayList<>();
        for (String sev : severities) {
            long count;
            if ("Unassessed".equals(sev)) {
                count = reportRepository.countTotal() - reportRepository.countBySeverity("Low")
                      - reportRepository.countBySeverity("Moderate")
                      - reportRepository.countBySeverity("High")
                      - reportRepository.countBySeverity("Critical");
                if (count < 0) count = 0;
            } else {
                count = reportRepository.countBySeverity(sev);
            }
            double pct = Math.round((count * 100.0 / total) * 100.0) / 100.0;
            result.add(SeverityDistributionItem.builder()
                .severity(sev).count(count).percentage(pct).build());
        }
        return result;
    }

    @Transactional(readOnly = true)
    public List<CategoryDistributionItem> categoryDistribution() {
        List<InfrastructureType> all = infraTypeRepository.findAll();
        List<Report> allReports = reportRepository.findAll();
        List<CategoryDistributionItem> result = new ArrayList<>();
        for (InfrastructureType it : all) {
            long count = allReports.stream().filter(r -> r.getInfrastructureType() != null
                && r.getInfrastructureType().getId().equals(it.getId())).count();
            long crit = allReports.stream().filter(r -> r.getInfrastructureType() != null
                && r.getInfrastructureType().getId().equals(it.getId())
                && ("Critical".equals(r.getAiSeverity()) || "Critical".equals(r.getFinalSeverity()))).count();
            result.add(CategoryDistributionItem.builder()
                .category(it.getName()).count(count).criticalCount(crit).build());
        }
        return result;
    }

    @Transactional(readOnly = true)
    public List<MonthlyTrendItem> monthlyTrend(int months) {
        LocalDateTime since = LocalDateTime.now().minusMonths(months);
        List<Report> reports = reportRepository.findAll().stream()
            .filter(r -> r.getCreatedAt() != null && r.getCreatedAt().isAfter(since))
            .collect(Collectors.toList());

        Map<String, long[]> byMonth = new TreeMap<>();  // [reports, resolved]
        for (Report r : reports) {
            String month = r.getCreatedAt().format(DateTimeFormatter.ofPattern("yyyy-MM"));
            long[] arr = byMonth.computeIfAbsent(month, k -> new long[2]);
            arr[0]++;
            if ("Resolved".equals(r.getStatus())) arr[1]++;
        }

        return byMonth.entrySet().stream()
            .map(e -> MonthlyTrendItem.builder()
                .month(e.getKey())
                .reports(e.getValue()[0])
                .resolved(e.getValue()[1])
                .build())
            .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<DistrictAnalyticsItem> districtAnalytics() {
        List<District> all = districtRepository.findAll();
        List<Report> allReports = reportRepository.findAll();
        List<DistrictAnalyticsItem> result = new ArrayList<>();
        for (District d : all) {
            List<Report> reports = allReports.stream()
                .filter(r -> r.getDistrict() != null && r.getDistrict().getId().equals(d.getId()))
                .collect(Collectors.toList());
            long crit = reports.stream().filter(r -> "Critical".equals(r.getAiSeverity())
                || "Critical".equals(r.getFinalSeverity())).count();
            long resolved = reports.stream().filter(r -> "Resolved".equals(r.getStatus())).count();
            double avgP = reports.stream()
                .mapToDouble(r -> priorityScoreRepository.findFirstByReportIdOrderByCreatedAtDesc(r.getId())
                    .map(PriorityScore::getScore).orElse(0.0))
                .average().orElse(0.0);
            result.add(DistrictAnalyticsItem.builder()
                .district(d.getName())
                .reports(reports.size())
                .critical(crit)
                .resolved(resolved)
                .avgPriority(Math.round(avgP * 100.0) / 100.0)
                .build());
        }
        return result;
    }

    @Transactional(readOnly = true)
    public Map<String, Object> responseTime() {
        Map<String, Object> map = new HashMap<>();
        Double avg = reportRepository.avgResponseTimeHours();
        Double min = reportRepository.minResponseTimeHours();
        Double max = reportRepository.maxResponseTimeHours();
        map.put("avg_hours", avg != null ? Math.round(avg * 100.0) / 100.0 : null);
        map.put("min_hours", min != null ? Math.round(min * 100.0) / 100.0 : null);
        map.put("max_hours", max != null ? Math.round(max * 100.0) / 100.0 : null);
        return map;
    }

    @Transactional(readOnly = true)
    public Map<String, Object> citizenParticipation() {
        Map<String, Object> map = new HashMap<>();
        map.put("total_citizens", userRepository.count());
        map.put("citizens_who_reported", userRepository.countActiveReporters());
        map.put("citizens_who_verified", userRepository.countVerifiers());
        Double avgVerifs = reportRepository.avgVerificationsPerReport();
        map.put("avg_verifications_per_report", avgVerifs != null
            ? Math.round(avgVerifs * 100.0) / 100.0 : 0.0);
        return map;
    }

    @Transactional(readOnly = true)
    public Map<String, Object> repeatIncidents(double thresholdKm) {
        List<Report> reports = reportRepository.findAllForClustering();
        List<Map<String, Object>> clusters = new ArrayList<>();
        Set<Long> used = new HashSet<>();
        for (int i = 0; i < reports.size(); i++) {
            Report r1 = reports.get(i);
            if (used.contains(r1.getId())) continue;
            List<Long> groupIds = new ArrayList<>();
            groupIds.add(r1.getId());
            for (int j = i + 1; j < reports.size(); j++) {
                Report r2 = reports.get(j);
                if (used.contains(r2.getId())) continue;
                if (r1.getInfrastructureType() == null || r2.getInfrastructureType() == null) continue;
                if (!r1.getInfrastructureType().getId().equals(r2.getInfrastructureType().getId())) continue;
                double d = com.infraguard.service.GeospatialService.haversineKm(
                    r1.getLatitude(), r1.getLongitude(),
                    r2.getLatitude(), r2.getLongitude()
                );
                if (d <= thresholdKm) {
                    groupIds.add(r2.getId());
                    used.add(r2.getId());
                }
            }
            if (groupIds.size() > 1) {
                Map<String, Object> cluster = new HashMap<>();
                cluster.put("infrastructure_type_id", r1.getInfrastructureType().getId());
                Map<String, Double> center = new HashMap<>();
                center.put("lat", r1.getLatitude());
                center.put("lng", r1.getLongitude());
                cluster.put("center", center);
                cluster.put("count", groupIds.size());
                cluster.put("report_ids", groupIds);
                clusters.add(cluster);
                used.add(r1.getId());
            }
        }
        return Map.of("clusters", clusters);
    }
}
